objtuple: Allow tuple-compatible classes like namedtuple to call mp_obj_tuple_get

jepler opened the pull request:
.. adding a private macro mp_obj_is_tuple_compatible to encapsulate the check, which is used in two locations.

Closes: #5005

jepler committed:
objtuple: Extract "compatible type" check to a macro

jepler committed:
objtuple: Allow compatible subclasses in mp_obj_tuple_get

Closes: #5005

dpgeorge commented on py/objtuple.c:
```diff
@@ -249,7 +252,7 @@ mp_obj_t mp_obj_new_tuple(size_t n, const mp_obj_t *items) {
 
 void mp_obj_tuple_get(mp_obj_t self_in, size_t *len, mp_obj_t **items) {
-    assert(mp_obj_is_type(self_in, &mp_type_tuple));
+    mp_check_self(mp_obj_is_tuple_compatible(self_in));
```
As you found out, `mp_obj_tuple_get` is really an (internal) C-API function.  So it is assumed that the caller knows what they are doing, and therefore there's an `assert` rather than an `mp_check_self`.  The latter is only used when incoming objects might be of the wrong type due to a user's script.

So this should really still be an `assert`.  Then the change in this PR just opens up the possibility that this C-API function can have tuple-compatible objects passed in.

jepler replied:
Done.  I also renamed the macro _mp_obj_is_tuple_compatible so future readers of the code will suspect it does not follow the usual pattern of mp_obj_is_TYPE.

jepler replied:
I also changed the other site from using mp_check_self to assert, as it appears to be "sufficiently internal" as well.

dpgeorge commented on py/objtuple.c:
```diff
@@ -31,6 +31,9 @@
 #include "py/runtime.h"
 
+// type check is done on getiter method to allow tuple, namedtuple, attrtuple
+#define mp_obj_is_tuple_compatible(o) (mp_obj_is_obj(o) && ((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type->getiter == mp_obj_tuple_getiter)
```
To retain existing behaviour/code size I think this should be exactly what was in `tuple_cmp_helper`, namely `mp_obj_get_type(o)->getiter == mp_obj_tuple_getiter`.

(Sorry, I was very rough before when I suggested to copy how objdict works.)

jepler committed:
objtuple: respond to review comments

 - is_tuple_compatible: Remove mp_obj_is_obj from check, to save code size

 - rename it so a future developer will not be deceived that it follows
   the general pattern of mp_obj_is_TYPE to check mp_obj_is_obj

 - Change both sites to assert, rather than mp_check_self, as they are
   internal APIs and user facing type checks should have been performed
   previously

This last is a potential behavioral change; in v1.9.4, tuple_cmp_helper
did use mp_check_self.  However, the same "is internal API" argument
appears to apply, since its caller mp_obj_tuple_binary_op has already
done `mp_obj_tuple_t *o = MP_OBJ_TO_PTR(lhs);` unchecked

dpgeorge commented on py/objtuple.c:
```diff
@@ -31,6 +31,10 @@
 
+// type check is done on getiter method to allow tuple, namedtuple, attrtuple
+// Assumes mp_obj_is_obj has already been checked
+#define _mp_obj_is_tuple_compatible(o) ((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type->getiter == mp_obj_tuple_getiter
```
It may not be that the object is a true object.  Eg it could be called as `mp_obj_tuple_get(MP_OBJ_NEW_SMALL_INT(1), ...)`.  Of course this is silly but the existing code will catch this (if assertions are enabled), which is exactly what the assertion is for.

dpgeorge commented on py/objtuple.c:
```diff
@@ -102,7 +106,7 @@ STATIC mp_obj_t mp_obj_tuple_make_new(const mp_obj_type_t *type_in, size_t n_arg
 STATIC mp_obj_t tuple_cmp_helper(mp_uint_t op, mp_obj_t self_in, mp_obj_t another_in) {
     // type check is done on getiter method to allow tuple, namedtuple, attrtuple
-    mp_check_self(mp_obj_get_type(self_in)->getiter == mp_obj_tuple_getiter);
+    assert(_mp_obj_is_tuple_compatible(self_in));
```
This should actually stay as a call to `mp_check_self(...)` because the incoming object can be supplied by a user from a script.  It may already be assumed to be a concrete object (ie a pointer) but that doesn't mean it can just be an assert here.

dpgeorge reviewed (commented):
Thanks for persisting with this.

I think the best option is to try and make the simplest change that changes only one thing.  I think that would be making a macro for `mp_obj_get_type(self_in)->getiter == mp_obj_tuple_getiter`, and using it in these 2 places.

jepler committed:
objtuple: respond to review comments again

mp_obj_is_tuple_compatible: use mp_obj_get_type as this is code size
neutral and removes a caveat.  remove underscore, because it functions
"as normal"

tuple_cmp_helper: restore behavior of checking type, not asserting it.
While this routine may appear to be an internal routine (STATIC
qualification), it actually is intended to be the site of type
checking.

jepler committed:
objtuple: remove redundant comment

jepler commented:
I've responded to your review requests again.  When we've reached a desirable endpoint, would you like me to rebase this and squash out the false starts?

dpgeorge commented on py/objtuple.c:
```diff
@@ -31,6 +31,9 @@
 #include "py/runtime.h"
 
+// type check is done on getiter method to allow tuple, namedtuple, attrtuple
+#define mp_obj_is_tuple_compatible(o) (mp_obj_get_type(o)->getiter == mp_obj_tuple_getiter
```
Missing a trailing `)`

dpgeorge commented:
Thanks for updating.  It looks good now (except for typo with missing `)`).

> would you like me to rebase this and squash out the false starts?

If you like.  Otherwise I'm happy to do this during merge.

dpgeorge commented:
I fixed the parenthesis, squashed and merged in 12f13ee6346d8fd029fc2ecec06d50b5f7f6b252

Thanks!

jepler commented:
Thank you!
